libsel4vm/src/arch/arm/vgic/vgic.c

```diff
@@ -194,24 +194,42 @@ struct irq_queue {
     size_t tail;
 };
 
+typedef struct vgic_core {
+    // Mirrors the vcpu list registers
+    struct virq_handle *lr_shadow[NUM_LIST_REGS];
+    // IRQs that would not fit in the vcpu list registers
+    struct irq_queue irq_queue;
+    // Complete set of virtual irqs
+    struct virq_handle *sgi_ppi_irq[NUM_SGI_VIRQS + NUM_PPI_VIRQS];
+} vgic_core_t;
+
 typedef struct vgic {
-/// Mirrors the vcpu list registers
-    struct virq_handle *lr_shadow[CONFIG_MAX_NUM_NODES][NUM_LIST_REGS];
-/// IRQs that would not fit in the vcpu list registers
-    struct irq_queue irq_queue[CONFIG_MAX_NUM_NODES];
-/// Complete set of virtual irqs
-    struct virq_handle *sgi_ppi_irq[CONFIG_MAX_NUM_NODES][NUM_SGI_VIRQS + NUM_PPI_VIRQS];
-    struct virq_handle *virqs[MAX_VIRQS];
-/// Virtual distributor registers
+    // Virtual distributor registers
     struct gic_dist_map *dist;
+    // global interrupts
+    struct virq_handle *virqs[MAX_VIRQS];
+    // core specific interrupt context
+    vgic_core_t core[CONFIG_MAX_NUM_NODES];
 } vgic_t;
 
 static struct vgic_dist_device *vgic_dist;
 
+static vgic_core_t *get_vgic_core(vgic_t *vgic, int vcpu_id)
+{
+    if (!vgic || (vcpu_id < 0) || (vcpu_id >= ARRAY_SIZE(vgic->core))) {
+        assert(0);
+        return NULL;
+    }
+
+    return &(vgic->core[vcpu_id]);
+}
+
 static struct virq_handle *virq_get_sgi_ppi(vgic_t *vgic, vm_vcpu_t *vcpu, int virq)
 {
-    assert(vcpu->vcpu_id < CONFIG_MAX_NUM_NODES);
-    return vgic->sgi_ppi_irq[vcpu->vcpu_id][virq];
+    vgic_core_t *vgic_core = get_vgic_core(vgic, vcpu->vcpu_id);
+    assert(vgic_core);
+    assert((virq >= 0) && (virq <= ARRAY_SIZE(vgic_core->sgi_ppi_irq)));
```

Comment:
Minor point: this should be `<` rather than `<=` I think.

Reply:
Thanks, you are right. Fixed.